crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -1233,6 +1246,59 @@ impl Room {
         Ok(())
     }
 
+    /// Removes the seen knock request ids that are no longer valid given the
+    /// current room members.
+    pub async fn remove_outdated_seen_knock_requests_ids(&self) -> StoreResult<()> {
+        let not_knocked_memberships = RoomMemberships::all().difference(RoomMemberships::KNOCK);
+        let not_knocked_user_ids = BTreeSet::from_iter(
+            self.store.get_user_ids(self.room_id(), not_knocked_memberships).await?,
+        );
+
+        let mut current_seen_events_guard = self.seen_knock_request_ids_map.write().await;
+        // We're not calling `get_seen_join_request_ids` here because we need to keep
+        // the Mutex's guard until we've updated the data
+        let mut current_seen_events = if current_seen_events_guard.is_none() {
+            self.load_cached_knock_request_ids().await?
+        } else {
+            current_seen_events_guard.clone().unwrap()
+        };
```

Comment:
Think this is the third duplication of this piece of code, can we refactor it into a small helper?